R/utils.R

```diff
@@ -259,85 +259,560 @@ rowcol_to_dim <- function(row, col) {
+  }
 
-    rows <- as.integer(args[[rows]])
-    cols <- col2int(args[[cols]])
+  rows_arg <- col2int(rows_arg)
+  cols_arg <- args$cols
+  x <- args$x
+  # rows_and_cols_present <- all(c("rows", "cols") %in% nams)
 
-  } else {
 
-    if (cnam_null) col_names <- exp_name
-    if (rnam_null) row_names <- FALSE
+  # Find column location id if `cols` is named.
+  if (x_has_colnames && !is.null(cols_arg)) {
+    is_cols_a_colname <- cols_arg %in% colnames(x)
 
-    assert_class(col_names, "logical")
-    assert_class(row_names, "logical")
+    if (any(is_cols_a_colname)) {
+      if (length(is_cols_a_colname) != 1) {
+        stop(
+          "Supplying multiple column names is not supported by the `wb_dims()` helper, use the `cols`  arguments instead.",
+          "\n Use a single `cols` at a time with `wb_dims()`"
+        )
+      }
+      # message("Transforming col name = '", cols_arg, "' to `cols = ", which(colnames(x) == cols_arg), "`")
+      cols_arg <- which(colnames(x) == cols_arg)
 
-    # wb_dims(data.frame())
-    x <- as.data.frame(x)
-    rows <- seq_len(nrow(x) + col_names)
-    cols <- seq_len(ncol(x) + row_names)
+    }
+  }
+
+  if (!is.null(rows_arg)) {
+    assert_class(rows_arg, class = "integer", arg_nm = "rows")
+  }
+
+  if (!is.null(cols_arg)) {
+    cols_arg <- col2int(cols_arg)
+    assert_class(cols_arg, class = "integer", arg_nm = "cols")
+  }
+
+  frow_null <- is.null(args$from_row)
+  srow <- args$from_row %||% 1L
+  srow <- as.integer(srow - 1L)
+
+
+  fcol_null <- is.null(args$from_col)
+  scol <- col2int(args$from_col) %||% 1L
+  scol <- scol - 1L
+  # after this point, no assertion, assuming all elements to be acceptable
+
+  # from_row / from_col = 0 only acceptable in certain cases.
+  if (!all(length(scol) == 1, length(srow) == 1)) {
+    stop("Internal error. At this point scol and srow should have length 1.")
+  }
+  if (!x_present && (identical(scol, -1L) || identical(srow, -1L))) {
+    stop("`from_row/col` = 0 only makes sense with `x` present")
+  }
+
+
+  # if `!x` return early
+  if (!x_present) {
+    row_span <- srow + rows_arg %||% 1L
+    col_span <- scol + cols_arg %||% 1L
+    if (identical(row_span, 0L)) {
+      stop("Providing `rows = 0` without an object with dimensions is not supported", "Use `rows = 1`.")
+    }
+    if (identical(col_span, 0L)) {
+      stop("Providing `cols = 0` without an object with dimensions is not supported", "Use `cols = 1`.")
+    }
+
+    if (length(row_span) == 1 && length(col_span) == 1) {
+      # A1
+      row_start <- row_span
+      col_start <- col_span
+      dims <- rowcol_to_dim(row_start, col_start)
+    } else {
+      # A1:B2
+      dims <- rowcol_to_dims(row_span, col_span)
+    }
+    return(dims)
+  }
 
+  # Fix an error I created
+  # if (frow_null)  {
+  #   srow <- srow + 1L
+  # }
+  # if (fcol_null) {
+  #   scol <- scol + 1L
+  # }
```

Comment:
commented code should be removed same with commented messages. either include them in some `if (debug) {}` statement or remove them entirely (my preference).